### PR TITLE
Added support for SCANNER_TRIVY_IGNORE_POLICY

### PR DESCRIPTION
**Pull request by pja-kit**
Signed-off-by: Peter Jakobsen <pja@kvalitetsit.dk>

This is a fix for: https://github.com/aquasecurity/harbor-scanner-trivy/issues/162

### CODE REVIEW

**Comment by danielpacak** [comment]
@pja-kit Thank you for this PR. Before I jump on code review I'd like to better understand a use case that you're trying to cover here.
When you use trivy with the `--ignore-policy` flag you specify the Rego file to evaluate each vulnerability in the specified container image.

However, this adapter service is run for multiple different images hosted by a Harbor registry. Do you want to use the same Rego file for each and every image? Notice that you'll have to restart the service to change Rego script.

**Comment by pja-kit** [comment]
> @pja-kit Thank you for this PR. Before I jump on code review I'd like to better understand a use case that you're trying to cover here.
> When you use trivy command with the `--ignore-policy` flag you specify the Rego file to evaluate each vulnerability in the specified container image.
> 
> However, this adapter service is run for multiple different images hosted in a Harbor registry. Do you want to use the same Rego file for each and every image?
> 
> Beyond that, there is a draft proposal to add Rego policies to Harbor directly to evaluate vulnerability reports returned by scanner adapters, so this change request won't be relevant anymore.

The background for this PR is that we need the posibility to filter which vulnerabilities we show in Harbor, so the draft proposal for Harbor would meet our requirements. However until this draft is implemented we would like to use the Trivy functionality to do this.

We are using a single rego file for every image in our Harbor repo, and this rego file is deployed as a config map mounted into the Pod running Trivy. 

From my point of view the Rego file i similar to the SCANNER_TRIVY_VULN_TYPE but with greater control over which vulnerabilities to return to Harbor.

If so required I can add support for the Rego-file in a ConfigMap in the Helm chart: https://github.com/aquasecurity/harbor-scanner-trivy/tree/main/helm/harbor-scanner-trivy

**Comment by danielpacak** [comment]
This is helpful @pja-kit In this case I'd vote for Helm chart update in scope of this PR. BTW, are you using the Helm chart from this repository or goharbor/harbor-helm?

**Comment by pja-kit** [comment]
> This is helpful @pja-kit In this case I'd vote for Helm chart update in scope of this PR. BTW, are you using the Helm chart from this repository or https://github.com/goharbor/harbor-helm?

@danielpacak I've added support for supplying the rego ignore policy in the Helm values.yaml. 
If scanner.trivy.ignorepolicy is set when the Helm chart is rendered a configmap is created and mounted into the Statefulset pods.

I've added an example rego file commented out in values.yaml.

File: README.md [hunk 1]
```diff
@@ -104,6 +104,7 @@ Configuration of the adapter is done via environment variables at startup.
 | `SCANNER_TRIVY_VULN_TYPE`                 | `os,library`                       | Comma-separated list of vulnerability types. Possible values are `os` and `library`. |
 | `SCANNER_TRIVY_SEVERITY`                  | `UNKNOWN,LOW,MEDIUM,HIGH,CRITICAL` | Comma-separated list of vulnerabilities severities to be displayed                   |
 | `SCANNER_TRIVY_IGNORE_UNFIXED`            | `false`                            | The flag to display only fixed vulnerabilities                                       |
+| `SCANNER_TRIVY_IGNORE_POLICY`             | ``                                 | The path for the Trivy Ignore policy file                                        |
```
Comment (danielpacak) [inline README.md:107]:
```suggestion
| `SCANNER_TRIVY_IGNORE_POLICY`             | ``                                 | The path for the Trivy ignore policy OPA Rego file                                        |
```

File: helm/harbor-scanner-trivy/README.md [hunk 2]
```diff
@@ -91,6 +91,7 @@ The following table lists the configurable parameters of the scanner adapter cha
 | `scanner.trivy.reportsDir`            | Trivy reports directory                                                 | `/home/scanner/.cache/reports` |
 | `scanner.trivy.debugMode`             | The flag to enable or disable Trivy debug mode                          | `false` |
 | `scanner.trivy.vulnType`              | Comma-separated list of vulnerability types. Possible values are `os` and `library`. | `os,library` |
+| `scanner.trivy.ignorepolicy`          | The OPA rego file which will be added to a ConfigMap and mapped into Trivy | `     ` |
```
Comment (danielpacak) [inline helm/harbor-scanner-trivy/README.md:94]:
```suggestion
| `scanner.trivy.ignorepolicy`          | The OPA rego script used by Trivy to evaluate each vulnerability | `     ` |
```

File: pkg/etc/config.go [hunk 3]
```diff
@@ -30,6 +30,7 @@ type Trivy struct {
 	VulnType      string `env:"SCANNER_TRIVY_VULN_TYPE" envDefault:"os,library"`
 	Severity      string `env:"SCANNER_TRIVY_SEVERITY" envDefault:"UNKNOWN,LOW,MEDIUM,HIGH,CRITICAL"`
 	IgnoreUnfixed bool   `env:"SCANNER_TRIVY_IGNORE_UNFIXED" envDefault:"false"`
+	IgnorePolicy  string `env:"SCANNER_TRIVY_IGNORE_POLICY" envDefault: ""`
```
Comment (danielpacak) [inline pkg/etc/config.go:33]:
```suggestion
	IgnorePolicy  string `env:"SCANNER_TRIVY_IGNORE_POLICY"`
```

File: pkg/trivy/wrapper.go [hunk 4]
```diff
@@ -121,6 +121,11 @@ func (w *wrapper) prepareScanCmd(imageRef ImageRef, outputFile string) (*exec.Cm
 		args = append([]string{"--skip-update"}, args...)
 	}
 
+	if w.config.IgnorePolicy != "" {
+		args = append([]string{"--ignore-policy",w.config.IgnorePolicy}, args...)
```
Comment (danielpacak) [inline pkg/trivy/wrapper.go:125]: Use gofmt to fix formatting

**Review by danielpacak — CHANGES_REQUESTED** [review]
Great job @pja-kit I test in my env and it works like a charm. I just left a few minor change requests before we can merge your PR.

File: pkg/trivy/wrapper.go (hunk 4 above)
Comment (pja-kit) [reply pkg/trivy/wrapper.go]: Hi @danielpacak 
I've done your requested changed to the PR.

**Review by danielpacak — APPROVED** [review]

**Pull request merged**